src/olympia/addons/tasks.py

```diff
@@ -78,6 +84,56 @@ def delete_preview_files(id, **kw):
     Preview.delete_preview_files(sender=None, instance=Preview(id=id))
 
 
+@task
+@use_primary_db
+def delete_all_addon_media_with_backup(id, **kwargs):
+    """Delete all media files for a given add-on by id, backing them up to
+    reported content bucket in cloud storage.
+
+    Used when force-disabling an add-on."""
+    addon = Addon.unfiltered.all().no_transforms().get(pk=id)
+    previews = list(Preview.objects.filter(addon__id=id).values_list('id', flat=True))
+    version_previews = list(
+        VersionPreview.objects.filter(version__addon__id=id).values_list(
+            'id', flat=True
+        )
+    )
+    if addon.icon_type and (ext := mimetypes.guess_extension(addon.icon_type)):
+        base_icon_path = os.path.join(addon.get_icon_dir(), str(addon.pk))
+        icon_path = f'{base_icon_path}-original{ext}'
+        if storage.exists(icon_path):
+            backup_file_name = copy_file_to_backup_storage(icon_path, addon.icon_type)
+            addon.update(icon_hash=os.path.splitext(backup_file_name)[0])
+        remove_icons(base_icon_path)
+    for preview in previews:
+        if not storage.exists(preview.original_path):
+            continue
+        backup_file_name = copy_file_to_backup_storage(
+            preview.original_path, mimetypes.guess_type(preview.original_path)[0]
+        )
+        preview.update(image_hash=os.path.splitext(backup_file_name)[0])
```

Comment:
This is probably the most controversial part. We have to store the backup name somewhere. I've chosen to keep an `image_hash` for all previews, like we do for icons, but also to re-use it, overwriting it with the backup file name (which is also a hash, except it's built from the filepath as well) so that we can use it for restoring.

Is it too weird ? The alternative would be to introduce a new model to track all `Preview` and keep a hash for it, which feels a bit annoying given that the `Preview` instances themselves are not deleted.

Reply:
yeah... it _is_ a bit weird to be hashing two entirely different things and storing them in the same field.